etl/analytics/common.py

```diff
@@ -1,63 +1,315 @@
+    sql : str
+        SQL query to execute.
+    datasette_csv_url : str
+        URL of the Datasette CSV endpoint.
+    chunk_size : int
+        Number of rows to fetch in each chunk.
+        The default is 10,000 (which is the maximum number of rows that Datasette can return in a single request).
+        If the query contains a LIMIT clause, it should be smaller than chunk_size (otherwise, an error is raised).
+        If the query does not contain a LIMIT clause, the query is paginated using LIMIT and OFFSET.
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame containing the results of the query.
+
+    """
+    # Check if the query contains a LIMIT clause.
+    limit_match = re.search(r"\bLIMIT\s+(\d+)(?:\s+OFFSET\s+(\d+))?\b", sql, re.IGNORECASE)
+
+    # Clean the query.
+    sql_clean = clean_sql_query(sql)
+
+    if limit_match:
+        # If a LIMIT clause already exists, check if it's larger than the limit.
+        limit_value = int(limit_match.group(1))
+        if limit_value > MAX_DATASETTE_N_ROWS:
+            raise ValueError(
+                f"Query LIMIT ({limit_value}) exceeds Datasette's maximum row limit ({MAX_DATASETTE_N_ROWS}). Either use a lower value for the limit, or set no limit (and pagination will be used)."
+            )
+        else:
+            # Given that there is a LIMIT clause, and the value is small, execute the query as-is.
+            full_url = f"{datasette_csv_url}?" + urllib.parse.urlencode({"sql": sql_clean, "_size": "max"})
+            # Fetch data as a dataframe, or raise an error (e.g. if query is too long).
+            df = _try_to_execute_datasette_query(sql_url=full_url, warn=True)
+    else:
+        # If there is no LIMIT clause, paginate using LIMIT/OFFSET.
+        offset = 0
+        dfs = []
+        while True:
+            # Prepare query for this chunk.
+            full_url = f"{datasette_csv_url}?" + urllib.parse.urlencode(
+                {"sql": f"{sql_clean} LIMIT {chunk_size} OFFSET {offset}", "_size": "max"}
+            )
+            # Fetch data for current chunk.
+            df_chunk = _try_to_execute_datasette_query(sql_url=full_url)
+            if len(df_chunk) == chunk_size:
+                # Add data for current chunk to the list.
+                dfs.append(df_chunk)
+                # Update offset.
+                offset += chunk_size
+            else:
+                # If fewer rows than the maximum (or even zero rows) are fetched, this must be the last chunk.
+                dfs.append(df_chunk)
+                break
+
+        if len(dfs) == 0:
+            # If no data was fetched, return an empty dataframe.
+            df = pd.DataFrame()
+        else:
+            # Concatenate all chunks of data.
+            df: pd.DataFrame = pd.concat(dfs, ignore_index=True)  # type: ignore
+
+    return df
+
+
+def read_metabase(sql: str) -> pd.DataFrame:
+    """Retrieve data from the Metabase API using an arbitrary sql query.
```

Comment:
maybe worth having a dedicated module, e.g. `analytics.metabase`, where you can define all metabase-related tools?